Initial InstantOn tests for WebProfile Transactions

dazavala opened the pull request:
For issue #21198

InstantOn tests for Jakarta Transactions support in web profiles 8.0, 9.x, and 10.0. This PR provides the following tests in a new build project, `io.openliberty.checkpoint_fat_transactions`:

- ServletTest: Essentially a transactions BVT. Verify transaction manager service composition, access to all transaction mgt API, and `UserTransaction` support in servlets when a server is restored after checkpoint at applications. The test uses a datasource to access an database constructed in memory by embedded derby.  The datasource is verified to be configurable at restore using properties in `server.env`.

- ServletStartupTest: Verify checkpoint/restore for a servlet that begins a transaction during application startup (`@WebServlet loadOnStartup=1`, `UserTransaction`)
 
- StartupBeanTest: Verify checkpoint/restore for a startup bean (EJB session bean) that begins a transaction during application startup (`@Startup`, `@TransactionManaged` (implied), `@TransactionAttribute=REQUIRES_NEW`, `UserTransaction`) 
 
- TransactionalBeanTest: Verify all `@Transactional` semantics within CDI managed beans annotated at class and method targets when a server is restored after checkpoint at applications.  The test also verifies a transactional `@Stereotype` and CDI `@Observer` methods.  The `@Transactional` annotation declares transaction semantics for business methods, only; not lifecycle methods.  There's no need to verify checkpoint failure behaviors during application startup.

- TransactionScopedBeanTest: Verify `@TransactionScoped` CDI managed bean instances are available within transaction contexts spanning a servlet and EJBs over multiple threads when a server is restored after checkpoint at applications.

dazavala commented:
#build

dazavala commented:
#build
Rebased on release. Shouldn't be an issue as 99% of these changes are new artifacts, but building again to be safe.

dazavala commented:
Need to add TestThatAlwaysPasses to ensure instantOn transaction FAT runs in personal build.

tjwatson commented on dev/io.openliberty.checkpoint_fat_transaction/fat/src/io/openliberty/checkpoint/fat/ServletStartupTest.java:
```diff
@@ -0,0 +1,146 @@
+    static final String SERVLET_NAME = APP_NAME + "/StartupServlet";
+
+    @Server(SERVER_NAME)
+    @TestServlet(servlet = StartupServlet.class, contextRoot = APP_NAME)
```
This is only a nit, but it can be confusing to use @TestServlet annotation when the servlet class being used does not have any `@Test` methods.  For this test class all the `@Test` methods are contained here and not in the servlet class.  That makes me believe this `@TestServlet` annotation is not necessary.

dazavala replied:
It's not; I removed it.

tjwatson commented on dev/io.openliberty.checkpoint_fat_transaction/fat/src/io/openliberty/checkpoint/fat/ServletStartupTest.java:
```diff
@@ -0,0 +1,146 @@
+    static void stopServer() {
+        if (server.isStarted()) {
+            try {
+                server.stopServer("CWWKZ0017W");
```
Maybe a quick comment here on what this expected warning is.

dazavala replied:
The correct warning would be nice.  WTRN0017W indicates the server is unable to begin a nested tran. This is an expected warning of the Servlet and TransactionalBean (cdi) tests.  I've seen this warning in the server logs during test execution, so I'm surprised the FAT framework didn't complain w/o this change.

tjwatson commented on dev/io.openliberty.checkpoint_fat_transaction/fat/src/io/openliberty/checkpoint/fat/TransactionScopedBeanTest.java:
```diff
@@ -0,0 +1,183 @@
+    static final String SERVLET_NAME = APP_NAME + "/transactionscopedbean";
+
+    @Server(SERVER_NAME)
+    @TestServlets({ @TestServlet(servlet = TransactionScopedBeanServlet.class, contextRoot = APP_NAME),
```
Similar confusion for `@TestServlets`.  I'm not sure what is supposed to happen when we have `@Test` methods from the test class AND the test servlets.  Do both get run?

dazavala replied:
I see the same behavior either way when the test method names in the Test class match those in the servlets.  Test methods run once.  I removed the `@TestServlet` from Test classes that do not rely on simplicity to run test methods defined in a servlet.

tjwatson commented on dev/io.openliberty.checkpoint_fat_transaction/publish/servers/checkpointTransactionServlet/server.xml:
```diff
@@ -0,0 +1,67 @@
+    <sslDefault outboundSSLRef="mySSLConfig" />
+
+    <!-- Server SSL configuration -->
+    <ssl id="defaultSSLConfig" keyStoreRef="defaultKeyStore" trustStoreRef="defaultTrustStore" /> 
```
What feature pulls in ssl support?

dazavala replied:
None of the features pull in ssl-1.0.  Yanked the ssl config and local testing works.  I suspect it's there to facilitate copy-paste in other test servers in the TX FAT.

tjwatson reviewed (commented):
Have a `dev/io.openliberty.checkpoint_fat_transaction/publish/servers/checkpointTransactionScopedBean/.server.xml.swp` file that needs removed.

dazavala commented:
Penance for not stacking a commit...

Per review: 
- Removed  `.swp` file and verified no others were committed
- Removed `@TestServlet` from any Test class that does not require the test framework to invoke `@Test` methods in a servlet.
- Removed `<ssl>` configuration from server `checkpointTransactionServlet`.
- Fixed and documented warning WTRN0017W when ignoring the warning for server shutdown. Warning is now specified only for test servers that require it. 

Per chat with @tjwatson
- Modified `build.gradle` to publish the `txtest-x.0` features in target `addRequiredLibraries` to ensure the test does not fail in the event build task `:com.ibm.ws.transaction.test.util:assemble` has not yet executed. Just as the Transaction FATs do, the Checkpoint Transaction FAT does not remove these features after completion.  Doing so may cause Transaction FATs to fail as they expect the features are permanently published once project `com.ibm.ws.transaction.test.util` builds.
- Modified `ServletTest`, `SimpleServlet.java`, and `checkpointTransactionServlet/server.xml` to verify the configuration for the data source used in transactions is picked up upon restore.  The server configuration contains an invalid jndi name which is reconfigured upon restore to the `@Resource` name injected into the test servlet.  

Bonus:
- Modified `bnd.bnd` with fewer project dependencies and setup the derby client binaries for future tests.

dazavala commented:
#build
Build the review changes.  See previous comment for change listing.

tjwatson commented on dev/io.openliberty.checkpoint_fat_transaction/bnd.bnd:
```diff
@@ -0,0 +1,87 @@
+
+fat.project: true
+
+fat.test.container.images: kyleaure/db2-ssl:3.0
```
Did not think you used test containers here, nor db2.

dazavala replied:
I removed line 26 from `bnd.bnd`.  It's not necessary and likely problematic.

dazavala committed:
Initial InstantOn tests for WebProfile Transactions

dazavala commented:
#build
One more time.  Removed fat test container image for DB2 from `bnd.bnd` and removed a few vestigial lines.

dazavala commented:
#libby
